<?xml version='1.0'?>

<launch>
  
  <arg name="r_name" default="racecar"/>
  <arg name="vesc_ns" default="/vesc"/>
  <arg name="r_num" default="1"/>

  <arg name="max_vel" default="0.5"/>
  <arg name="max_acc" default="0.1"/>

  <param name="max_vel" value="$(arg max_vel)"/>

  <!-- Load joint controller configurations from YAML file to parameter server -->
  <rosparam file="$(find racecar_control)/config/racecar_control.yaml" command="load" ns="/$(arg r_name)"/>

  <!-- load the controllers -->
  <node name="controller_manager" pkg="controller_manager" type="spawner" respawn="false"
        output="screen" ns="/$(arg r_name)" args="left_rear_wheel_velocity_controller 
                                            right_rear_wheel_velocity_controller
                                            left_front_wheel_velocity_controller 
                                            right_front_wheel_velocity_controller
                                            left_steering_hinge_position_controller 
                                            right_steering_hinge_position_controller
                                            joint_state_controller"/>

  <!-- convert joint states to TF transforms for rviz, etc -->
  <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" respawn="false" output="screen">
    <remap from="/joint_states" to="/$(arg r_name)/joint_states" />
  </node>

  <!-- lane keeping node -->
  <node pkg="racecar_control" type="lane_keeping.py" name="lane_keeping" output="screen">
    <remap from="/vesc/odom" to="$(arg vesc_ns)/odom"/>
    <remap from="/camera/image_raw" to="/$(arg r_name)/camera/image_raw"/>
    <remap from="/drive_parameters" to="/$(arg r_name)/drive_parameters"/>
    <remap from="/desired_vel" to="/$(arg r_name)/desired_vel"/>
  </node>

  <!-- drive_param to servo node -->
  <node pkg="racecar_control" type="drive_param_to_servo.py" name="drive_param_to_servo" output="screen">
    <remap from="/vesc/ackermann_cmd_mux/input/teleop" to="$(arg vesc_ns)/ackermann_cmd_mux/input/teleop"/>
    <remap from="/drive_parameters" to="/$(arg r_name)/drive_parameters"/>
  </node>

  <!-- servo node -->
  <node pkg="racecar_control" type="servo_commands.py" name="servo_commands" output="screen">
    <remap from="/ackermann_cmd_mux/output" to="$(arg vesc_ns)/low_level/ackermann_cmd_mux/output"/>
    <remap from="/left_rear_wheel_velocity_controller/command" to="/$(arg r_name)/left_rear_wheel_velocity_controller/command"/>
    <remap from="/right_rear_wheel_velocity_controller/command" to="/$(arg r_name)/right_rear_wheel_velocity_controller/command"/>
    <remap from="/left_front_wheel_velocity_controller/command" to="/$(arg r_name)/left_front_wheel_velocity_controller/command"/>
    <remap from="/right_front_wheel_velocity_controller/command" to="/$(arg r_name)/right_front_wheel_velocity_controller/command"/>
    <remap from="/left_steering_hinge_position_controller/command" to="/$(arg r_name)/left_steering_hinge_position_controller/command"/>
    <remap from="/right_steering_hinge_position_controller/command" to="/$(arg r_name)/right_steering_hinge_position_controller/command"/>
  </node>

  <!-- Allow for Gazebo to broadcast odom -->
  <param name="robot_name" value="$(arg r_name)"/>
  <node pkg="racecar_gazebo" name="gazebo_odometry_node" type="gazebo_odometry.py">
    <remap from="/vesc/odom" to="$(arg vesc_ns)/odom"/>
  </node>

</launch>
